File: package/main/src/Object/mapKeys.ts
/**
 * Creates an object with the same values as the given object, but
 * with keys transformed by the provided function.
 *
 * @param object - The source object.
 * @param function_ - The function invoked per key. Receives (value, key).
 * @returns A new object with transformed keys.
 *
 * @example
 * ```typescript
 * mapKeys({ a: 1, b: 2 }, (_value, key) => key.toUpperCase());
 * // { A: 1, B: 2 }
 * ```
 */
export const mapKeys = <T extends Record<string, unknown>>(
  object: T,
  function_: (value: T[keyof T], key: string) => string,
): Record<string, T[keyof T]> => {
  const result: Record<string, T[keyof T]> = {};
  const keys = Object.keys(object);
  const length = keys.length;
  let index = 0;

  while (index < length) {
    const key = keys[index];
    // Skip prototype pollution keys from source
    if (key === "__proto__" || key === "constructor" || key === "prototype") {
      index += 1;
      continue;
    }
    const value = object[key] as T[keyof T];
    const newKey = function_(value, key);
    // Prevent prototype pollution via transformed keys
    if (
      newKey === "__proto__" ||
      newKey === "constructor" ||
      newKey === "prototype"
    ) {
      index += 1;
      continue;
    }
    result[newKey] = value;
    index += 1;
  }

  return result;
};

File: package/umt_rust/src/object/map_keys.rs
use std::collections::HashMap;

use super::Value;

/// Creates an object with the same values as the given object, but
/// with keys transformed by the provided function.
///
/// # Arguments
///
/// * `object` - The source object.
/// * `function` - The function invoked per key. Receives (value, key).
///
/// # Returns
///
/// A new object with transformed keys.
///
/// # Example
///
/// ```rust
/// use std::collections::HashMap;
/// use umt_rust::object::Value;
/// use umt_rust::object::map_keys::umt_map_keys;
///
/// let mut map = HashMap::new();
/// map.insert("a".to_string(), Value::Int(1));
/// map.insert("b".to_string(), Value::Int(2));
///
/// let result = umt_map_keys(&map, |_, key| key.to_uppercase());
///
/// assert_eq!(result.get("A"), Some(&Value::Int(1)));
/// assert_eq!(result.get("B"), Some(&Value::Int(2)));
/// ```
pub fn umt_map_keys<F>(object: &HashMap<String, Value>, function: F) -> HashMap<String, Value>
where
    F: Fn(&Value, &str) -> String,
{
    let mut result = HashMap::new();

    for (key, value) in object.iter() {
        result.insert(function(value, key), value.clone());
    }

    result
}
